README
gtk vector screenshot module
============================

© 2011 Joachim Breitner <mail@joachim-breitner.de>

This gtk module allows you to take a screenshot of a running gtk-3 application
as a vector image, with fully scalable graphics and selectable text. It
supports rendering the application to PDF, SVG and PostScript.

Installation
============

To install the module, use the usual sequence:

./configure --prefix=/usr
make
sudo make install

And then ensure that the environment variable GTK_MODULES includes
gtk-vector-screenshot, e.g. in ~/.xsessionrc or by putting the file
52load-gtk-vector-screenshot-gtk-module into /etc/X11/Xsession.d.

If you do not install into /usr, you will have set GTK_PATH or pass the full
path of gtk-vector-screencast.so via GTK_MODULES when invoking the application.

Usage
=====

You need to ensure that the gtk-vector-screenshot module is loaded by the
application, see section Installation. Distribution packages should take care
of this, but you might need to login again after you install the package.

Then run take-vector-screenshot, press the button, click on the window and then
you can select the filename to save the image under.

You can watch a screencast of the program in action at
http://www.joachim-breitner.de/blog/archives/494-Better-PDF-screenshots-with-gtk-3.html

Protocol
========

The module is loaded by every application, and only activated using the
take-vector-screenshot program. The protocol is simple and based on X window
properties and X client messages: The gtk module sets the GTK_VECTOR_SCREENSHOT
property for each top-level X window to some arbitrary value (currently
"supported"). The control application sends an XClientMessage event to that
window of type GTK_VECTOR_SCREENSHOT and with event.window set to the receiving
window.

Development
===========

The latest release can be found on 
http://www.joachim-breitner.de/archive/gtk-vector-screenshot/

For the list of changes, see ChangeLog.

The code is hosted on GitHub:
https://github.com/nomeata/gtk-vector-screenshot

You can pull from the repository via
git clone https://github.com/nomeata/gtk-vector-screenshot.git
or if you are registered with gitorious, via
git@github.com:nomeata/gtk-vector-screenshot.git

Feel free to submit patches, but also feel free to contact me first with your
ideas, in case I have different plans.

BUGS and TODOs
==============
 * It only works with gtk3 applications.
 * The screenshot is taken the moment you press Save in the file chooser
   dialogue, not when you press the “Take vector screenshot” button.
 * Make the preview widget behave more sensible with regard to resizes.
 * Less hakish setting of the GTK_VECTOR_SCREENSHOT property.
 * Fallback code without IPC for windows, as for version 0.2.
 * Use GtkInvisible for the grab in take-vector-screenshot.c, as that is what
   its meant for, it seems.


ChangeLog
gtk-vector-screenshot (0.3.3)

  * Drop the noop gtk2 module (used to quench a warning about missing gtk
    module), set GTK3_MODULES instead.

 -- Joachim Breitner <mail@joachim-breitner.de>  Sat, 22 Jun 2013 22:13:25 +0200

gtk-vector-screenshot (0.3.2.1)

  * Fix linking problem

 -- Joachim Breitner <mail@joachim-breitner.de>  Sat, 22 Jun 2013 22:13:25 +0200

gtk-vector-screenshot (0.3.2)

  * The BadWindow error situation was futher improved by Michel Dänzer.

 -- Joachim Breitner <mail@joachim-breitner.de>  Mon, 17 Oct 2011 19:37:46 +0200

gtk-vector-screenshot (0.3.1)

  * Petr Gajdůšek <gajdusek.petr@centrum.cz> reported, analyized and patched
    an issue causing gtk application to crash with BadWindow errors when the
    gtk-vector-screenshot module is loaded.

 -- Joachim Breitner <mail@joachim-breitner.de>  Sat, 27 Aug 2011 10:55:43 +0200

gtk-vector-screenshot (0.3)

  * Uses code from xwininfo to reliably find the right X window, even with
    window managers such as metacity.
  * Smaller Save-As dialogue even with large application windows.
    Help getting this dialogue right with regard to sizing and resizing would
    be appreciated!

 -- Joachim Breitner <mail@joachim-breitner.de>  Sun, 10 Jul 2011 12:18:18 +0200

gtk-vector-screenshot (0.2)

  * Now with separate control application take-vector-screenshot and
    unobstrusive module that can be loaded permanently.

 -- Joachim Breitner <mail@joachim-breitner.de>  Fri, 24 Jun 2011 16:06:13 +0200

gtk-vector-screenshot (0.1)

  * Initial release.

 -- Joachim Breitner <mail@joachim-breitner.de>  Mon, 20 Jun 2011 14:50:42 +0200


module.c

// For strchrnul
#define _GNU_SOURCE

#include <math.h>
#include <string.h>
#include <libgen.h>

#include <gtk/gtk.h>

#include <cairo-pdf.h>
#include <cairo-svg.h>
#include <cairo-ps.h>

void pdfscreenshot_type_selected(GtkComboBox *format_combo, GtkFileChooser *chooser) {
    const char *the_id = gtk_combo_box_get_active_id(GTK_COMBO_BOX(format_combo));

    // Update filter
    GtkFileFilter *filter = gtk_file_filter_new ();
    char *pattern = g_strdup_printf("*.%s", the_id);
    gtk_file_filter_add_pattern (filter, pattern);
    gtk_file_chooser_set_filter (GTK_FILE_CHOOSER(chooser), filter);

    // Update filename extension
    gchar *filename = gtk_file_chooser_get_filename(chooser);
    if (filename != NULL)  {
        // strup result from basename as we are going to modify it.
        char *base = g_strdup(basename(filename));
        // Remove extension from string, if there is one.
        *(strchrnul(base, '.')) = '\0';
        
        char *new_filename = g_strdup_printf("%s.%s", base, the_id);
        gtk_file_chooser_set_current_name(chooser, new_filename);
        g_free(filename);
        g_free(base);
        g_free(new_filename);
    }
}

void pdfscreenshot_draw_preview (GtkWidget *widget, cairo_t *cr, gpointer window) {

    int draw_width = gtk_widget_get_allocated_width (widget);
    int draw_height = gtk_widget_get_allocated_height (widget);
    int win_width = gtk_widget_get_allocated_width (window);
    int win_height = gtk_widget_get_allocated_height (window);

    double scale = fmin(1, fmin (1.0*draw_width / win_width, 1.0*draw_height / win_height));

    cairo_scale(cr, scale, scale);

    cairo_translate(cr, (draw_width  - scale * win_width)/2,
                        (draw_height - scale * win_height)/2);

    gtk_widget_draw(window, cr);
}

void pdfscreenshot_draw_to_vector (GtkWidget *widget, const gchar* filename, cairo_surface_t * create_surface(const char *, double , double )) {
    cairo_surface_t *surface = 
            create_surface(filename,
                    1.0*gtk_widget_get_allocated_width (widget),
                    1.0*gtk_widget_get_allocated_height (widget));
    cairo_t *cr  = cairo_create(surface);
    gtk_widget_draw(widget, cr);
    cairo_destroy(cr);
    cairo_surface_destroy(surface);
}

void pdfscreenshot_draw_to_png (GtkWidget *widget, const gchar* filename) {
    cairo_surface_t *surface = cairo_image_surface_create (
                    CAIRO_FORMAT_ARGB32,
                    gtk_widget_get_allocated_width (widget),
                    gtk_widget_get_allocated_height (widget));
    cairo_t *cr  = cairo_create(surface);
    gtk_widget_draw(widget, cr);
    cairo_destroy(cr);
    cairo_surface_write_to_png (surface, filename);
    cairo_surface_destroy(surface);
}

void
pdfscreenshot_take_shot (GtkButton *button, gpointer our_window) {
    GList *toplevels = gtk_window_list_toplevels();

    for (GList *iter = toplevels; iter; iter = g_list_next(iter)) {
        GtkWindow *window = GTK_WINDOW(iter->data);
        const gchar *title = gtk_window_get_title(window);
        if (our_window != window && title != NULL) {
            // This seems to be the "other" window
            g_object_ref(window);

            GtkWidget *chooser = gtk_file_chooser_dialog_new (
                "Save vector screenshot",
                our_window,
                GTK_FILE_CHOOSER_ACTION_SAVE,
                GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
                GTK_STOCK_SAVE, GTK_RESPONSE_ACCEPT,
                NULL);

            char *filename = g_strdup_printf("%s.pdf", g_get_application_name());
            gtk_file_chooser_set_current_name (GTK_FILE_CHOOSER (chooser), filename);
            g_free(filename);


            gtk_window_set_transient_for (GTK_WINDOW(chooser), GTK_WINDOW(our_window));
            gtk_file_chooser_set_do_overwrite_confirmation (GTK_FILE_CHOOSER (chooser), TRUE);

            GtkWidget *format_combo = gtk_combo_box_text_new ();
            gtk_combo_box_text_append(GTK_COMBO_BOX_TEXT(format_combo)
                                     , "pdf" ,"Save as PDF (*.pdf)");
            gtk_combo_box_text_append(GTK_COMBO_BOX_TEXT(format_combo)
                                     , "svg" ,"Save as SVG (*.svg)");
            gtk_combo_box_text_append(GTK_COMBO_BOX_TEXT(format_combo)
                                     , "ps" ,"Save as PostScript (*.ps)");
            gtk_combo_box_text_append(GTK_COMBO_BOX_TEXT(format_combo)
                                     , "png" ,"Save as PNG (*.png)");
            gtk_combo_box_set_active(GTK_COMBO_BOX(format_combo),0);
            pdfscreenshot_type_selected(GTK_COMBO_BOX(format_combo), GTK_FILE_CHOOSER(chooser));
	    g_signal_connect(GTK_COMBO_BOX(format_combo),"changed",
	    	G_CALLBACK(pdfscreenshot_type_selected), chooser);


            GtkWidget *drawing_area = gtk_drawing_area_new ();
            g_signal_connect (G_OBJECT (drawing_area), "draw",
                    G_CALLBACK (pdfscreenshot_draw_preview), window);
            gtk_widget_set_size_request (drawing_area, 300, 300);
        
            GtkWidget *frame = gtk_frame_new("Preview");
            gtk_container_add(GTK_CONTAINER(frame), drawing_area);

	    GtkWidget *vbox = gtk_vbox_new(FALSE, 8);
            gtk_box_pack_start(GTK_BOX(vbox), format_combo, FALSE, FALSE, 0);
            gtk_box_pack_start(GTK_BOX(vbox), frame, TRUE, TRUE, 0);
            gtk_widget_show_all(vbox);

            gtk_file_chooser_set_extra_widget(GTK_FILE_CHOOSER(chooser), vbox);

            if (gtk_dialog_run (GTK_DIALOG(chooser)) == GTK_RESPONSE_ACCEPT) {
                char *filename = gtk_file_chooser_get_filename (GTK_FILE_CHOOSER (chooser));
		//GtkFileFilter *the_filter = gtk_file_chooser_get_filter(GTK_FILE_CHOOSER(chooser));
                const char *the_id = gtk_combo_box_get_active_id(GTK_COMBO_BOX(format_combo));

		if (!g_strcmp0(the_id,"pdf"))
                    pdfscreenshot_draw_to_vector(GTK_WIDGET(window),filename,
                            cairo_pdf_surface_create);
		else if (!g_strcmp0(the_id,"svg"))
                    pdfscreenshot_draw_to_vector(GTK_WIDGET(window),filename,
                            cairo_svg_surface_create);
		else if (!g_strcmp0(the_id,"ps"))
                    pdfscreenshot_draw_to_vector(GTK_WIDGET(window),filename,
                            cairo_ps_surface_create);
		else if (!g_strcmp0(the_id,"png"))
                    pdfscreenshot_draw_to_png(GTK_WIDGET(window),filename);
		else
                    printf("Unknown id \"%s\"\n", the_id);

                g_free(filename);                        
            }

            gtk_widget_destroy (chooser);
            g_object_unref(window);
            g_list_free(toplevels);
            return;
        }
    }

    GtkWidget *dialog = gtk_message_dialog_new (our_window,
         GTK_DIALOG_DESTROY_WITH_PARENT,
         GTK_MESSAGE_ERROR,
         GTK_BUTTONS_CLOSE,
         "No main window found.");
    gtk_dialog_run (GTK_DIALOG (dialog));
    gtk_widget_destroy (dialog);
}

void
pdfscreenshot_window_create()
{

    GtkWidget *icon = gtk_image_new_from_icon_name ("camera",GTK_ICON_SIZE_BUTTON);
    gtk_image_set_pixel_size(GTK_IMAGE(icon), 128);

    GtkWidget *button = gtk_button_new_with_label("Take vector screenshot...");
    gtk_button_set_image (GTK_BUTTON(button), icon);
    gtk_button_set_image_position (GTK_BUTTON(button), GTK_POS_TOP);

    GtkWidget *window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_window_set_title(GTK_WINDOW(window),"Vector screenshot taker");
    gtk_window_set_resizable(GTK_WINDOW(window), FALSE);
    gtk_window_set_keep_above(GTK_WINDOW(window),TRUE);
    gtk_container_add(GTK_CONTAINER(window), button);

    g_signal_connect(G_OBJECT(button), "clicked",
        G_CALLBACK(pdfscreenshot_take_shot), window);


    gtk_widget_show_all(GTK_WIDGET(window));
}

int
gtk_module_init(gint argc, char *argv[])
{
    pdfscreenshot_window_create();
    return FALSE;
}




gtk-vector-screenshot.c
/*
 * © 2011 Joachim Breitner <mail@joachim-breitner.de>
 *
 *   This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */


// For strchrnul
#define _GNU_SOURCE

#include <math.h>
#include <string.h>
#include <libgen.h>

#include <gtk/gtk.h>
#include <gdk/gdkx.h>

#include <cairo-pdf.h>
#include <cairo-svg.h>
#include <cairo-ps.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>

#ifndef MAX
#define MAX(a,b) (((a)>(b))?(a):(b))
#endif /* MAX */
#ifndef MIN
#define MIN(a,b) (((a)<(b))?(a):(b))
#endif /* MIN */

gchar *filename;
const gchar *type;

GdkAtom pdfscreenshot_atom;
char *supported_str = "supported";

XErrorHandler old_handler = (XErrorHandler) 0 ;

/*
 * This function handles the file type combo box callback in the Save As
 * dialogue. I would have expected that such functionality is already provided
 * by gtk...
 */
void pdfscreenshot_type_selected(GtkComboBox *format_combo, GtkFileChooser *chooser) {
    // The id is known to be the file extension.
    const char *the_id = gtk_combo_box_get_active_id(GTK_COMBO_BOX(format_combo));

    // Update filter, to only display appropriate files.
    GtkFileFilter *filter = gtk_file_filter_new ();
    char *pattern = g_strdup_printf("*.%s", the_id);
    gtk_file_filter_add_pattern (filter, pattern);
    gtk_file_chooser_set_filter (GTK_FILE_CHOOSER(chooser), filter);

    // Update filename extension in the chooser
    gchar *filename = gtk_file_chooser_get_filename(chooser);
    if (filename != NULL)  {
        // strdup result from basename as we are going to modify it.
        char *base = g_strdup(basename(filename));
        // Remove extension from string, if there is one.
        *(strchrnul(base, '.')) = '\0';
        // Assemble and set new suggested filename
        char *new_filename = g_strdup_printf("%s.%s", base, the_id);
        gtk_file_chooser_set_current_name(chooser, new_filename);
        g_free(filename);
        g_free(base);
        g_free(new_filename);
    }
}

/*
 * This function draws the main window in the preview pane of the Save As
 * dialogue. Not very useful, but we do it because we can.
 */
void pdfscreenshot_draw_preview (GtkWidget *widget, cairo_t *cr, gpointer window) {

    int draw_width = gtk_widget_get_allocated_width (widget);
    int draw_height = gtk_widget_get_allocated_height (widget);
    int win_width = gtk_widget_get_allocated_width (window);
    int win_height = gtk_widget_get_allocated_height (window);
    
    gtk_widget_set_size_request (widget, MIN(win_width,500), MIN(win_height,300));

    double scale = fmin(1, fmax (1.0*draw_width / win_width, 1.0*draw_height / win_height));

    cairo_scale(cr, scale, scale);

    cairo_translate(cr, (draw_width  - scale * win_width)/2,
                        (draw_height - scale * win_height)/2);

    gtk_widget_draw(window, cr);
}

/*
 * The main routine of saving a vector screenshot. Surprisingly simple. It is
 * parametized by the cairo surface creating function, as they all have the
 * same signature.
 */
void pdfscreenshot_draw_to_vector (GtkWidget *widget, const gchar* filename,
                    cairo_surface_t * create_surface(const char *, double , double )) {
    cairo_surface_t *surface = 
            create_surface(filename,
                    1.0*gtk_widget_get_allocated_width (widget),
                    1.0*gtk_widget_get_allocated_height (widget));
    cairo_t *cr  = cairo_create(surface);
    gtk_widget_draw(widget, cr);
    cairo_destroy(cr);
    cairo_surface_destroy(surface);
}

/*
 * Like above, but for PNG. Needs slightly different order of cairo function calls.
 */
void pdfscreenshot_draw_to_png (GtkWidget *widget, const gchar* filename) {
    cairo_surface_t *surface = cairo_image_surface_create (
                    CAIRO_FORMAT_ARGB32,
                    gtk_widget_get_allocated_width (widget),
                    gtk_widget_get_allocated_height (widget));
    cairo_t *cr  = cairo_create(surface);
    gtk_widget_draw(widget, cr);
    cairo_destroy(cr);
    cairo_surface_write_to_png (surface, filename);
    cairo_surface_destroy(surface);
}

/*
 * Called when the main button is pressed. Finds the main window, displays the
 * file chooser and saves the screenshot.
 */
void
pdfscreenshot_take_shot (GtkWindow *window) {
    // Set up the file chooser
    GtkWidget *chooser = gtk_file_chooser_dialog_new (
        "Save vector screenshot",
        window,
        GTK_FILE_CHOOSER_ACTION_SAVE,
        GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
        GTK_STOCK_SAVE, GTK_RESPONSE_ACCEPT,
        NULL);

    // Suggested file name derived from the application name.
    if (filename == NULL) {
        char *filename = g_strdup_printf("%s.pdf", g_get_application_name());
        gtk_file_chooser_set_current_name (GTK_FILE_CHOOSER (chooser), filename);
        g_free(filename);
    } else 
        gtk_file_chooser_set_filename (GTK_FILE_CHOOSER (chooser), filename);

    // Some generally useful setup for a Save As dialogue
    gtk_window_set_transient_for (GTK_WINDOW(chooser), GTK_WINDOW(window));
    gtk_file_chooser_set_do_overwrite_confirmation (GTK_FILE_CHOOSER (chooser), TRUE);

    // The combo box that selects the desired file type. We (ab)use the
    // id field for the file extension.
    GtkWidget *format_combo = gtk_combo_box_text_new ();
    gtk_combo_box_text_append(GTK_COMBO_BOX_TEXT(format_combo),
                             "pdf", "Save as PDF (*.pdf)");
    gtk_combo_box_text_append(GTK_COMBO_BOX_TEXT(format_combo),
                             "svg", "Save as SVG (*.svg)");
    gtk_combo_box_text_append(GTK_COMBO_BOX_TEXT(format_combo),
                             "ps",  "Save as PostScript (*.ps)");
    gtk_combo_box_text_append(GTK_COMBO_BOX_TEXT(format_combo),
                             "png", "Save as PNG (*.png)");
    // When this changes, we call pdfscreenshot_type_selected. 
    g_signal_connect(GTK_COMBO_BOX(format_combo),"changed",
        G_CALLBACK(pdfscreenshot_type_selected), chooser);
    // Default to PDF
    gtk_combo_box_set_active_id(GTK_COMBO_BOX(format_combo),type);

    // The framed drawing area is used for a preview of the window, as a gimmick.
    GtkWidget *drawing_area = gtk_drawing_area_new ();
    g_signal_connect (G_OBJECT (drawing_area), "draw",
            G_CALLBACK (pdfscreenshot_draw_preview), window);

    int win_width = gtk_widget_get_allocated_width (GTK_WIDGET(window));
    int win_height = gtk_widget_get_allocated_height (GTK_WIDGET(window));
    gtk_widget_set_size_request (drawing_area, MIN(win_width,500), MIN(win_height,300));

    GtkWidget *frame = gtk_aspect_frame_new("Preview",0.5,0,1,TRUE);
    gtk_container_add(GTK_CONTAINER(frame), drawing_area);

    // Shove both widgets in a vbox as the “extra” of the file chooser dialogue
    GtkWidget *vbox = gtk_vbox_new(FALSE, 8);
    gtk_box_pack_start(GTK_BOX(vbox), format_combo, FALSE, FALSE, 0);
    gtk_box_pack_start(GTK_BOX(vbox), frame, TRUE, TRUE, 0);
    gtk_widget_show_all(vbox);
    gtk_file_chooser_set_extra_widget(GTK_FILE_CHOOSER(chooser), vbox);

    // Run the dialogue and act if OK was pressed
    if (gtk_dialog_run (GTK_DIALOG(chooser)) == GTK_RESPONSE_ACCEPT) {
        // Read the filename and selected file type
        g_free(filename);
        filename = gtk_file_chooser_get_filename (GTK_FILE_CHOOSER (chooser));
        type = gtk_combo_box_get_active_id(GTK_COMBO_BOX(format_combo));

        // And save according to the selected file type.
        if (!g_strcmp0(type,"pdf"))
            pdfscreenshot_draw_to_vector(GTK_WIDGET(window),filename,
                    cairo_pdf_surface_create);
        else if (!g_strcmp0(type,"svg"))
            pdfscreenshot_draw_to_vector(GTK_WIDGET(window),filename,
                    cairo_svg_surface_create);
        else if (!g_strcmp0(type,"ps"))
            pdfscreenshot_draw_to_vector(GTK_WIDGET(window),filename,
                    cairo_ps_surface_create);
        else if (!g_strcmp0(type,"png"))
            pdfscreenshot_draw_to_png(GTK_WIDGET(window),filename);
        else
            // Should not happen.
            printf("Unknown id \"%s\"\n", type);
    }

    gtk_widget_destroy (chooser);
}

gboolean
pdfscreenshot_take_shot_soon (gpointer window) {
    pdfscreenshot_take_shot(GTK_WINDOW(window));
    return FALSE;
}

/*
 * Called when the main button is pressed. Finds the main window, and calls
 * pdfscreenshot_take_shot.
 */
void
pdfscreenshot_find_window (GtkButton *button, gpointer our_window) {
    GList *toplevels = gtk_window_list_toplevels();

    for (GList *iter = toplevels; iter; iter = g_list_next(iter)) {
        GtkWindow *window = GTK_WINDOW(iter->data);
        const gchar *title = gtk_window_get_title(window);
        if (our_window != window && title != NULL) {
            // This seems to be the "other" window that we want to shoot, so
            // grab hold of it
            g_object_ref(window);
            pdfscreenshot_take_shot(window);
            g_object_unref(window);
            g_list_free(toplevels);
            return;
        }
    }
    g_list_free(toplevels);

    // If we reach here, then we did not find a main window. Inform the user
    // about it.
    GtkWidget *dialog = gtk_message_dialog_new (our_window,
         GTK_DIALOG_DESTROY_WITH_PARENT,
         GTK_MESSAGE_ERROR,
         GTK_BUTTONS_CLOSE,
         "No main window found.");
    gtk_dialog_run (GTK_DIALOG (dialog));
    gtk_widget_destroy (dialog);
}

/*
 * The main window, with the one button.
 */
void
pdfscreenshot_window_create()
{
    // Is that icon always installed or do we have to ship it?
    GtkWidget *icon = gtk_image_new_from_icon_name ("camera",GTK_ICON_SIZE_BUTTON);
    gtk_image_set_pixel_size(GTK_IMAGE(icon), 128);

    GtkWidget *button = gtk_button_new_with_label("Take vector screenshot...");
    gtk_button_set_image (GTK_BUTTON(button), icon);
    gtk_button_set_image_position (GTK_BUTTON(button), GTK_POS_TOP);

    GtkWidget *window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_window_set_icon_name(GTK_WINDOW(window),"camera");
    gtk_window_set_title(GTK_WINDOW(window),"Vector screenshot taker");
    gtk_window_set_resizable(GTK_WINDOW(window), FALSE);
    gtk_window_set_keep_above(GTK_WINDOW(window),TRUE);
    gtk_container_add(GTK_CONTAINER(window), button);

    g_signal_connect(G_OBJECT(button), "clicked",
        G_CALLBACK(pdfscreenshot_find_window), window);

    gtk_widget_show_all(GTK_WIDGET(window));
}

/* 
 * Ignore all BadWindow errors.
 */
int
silent_error_handler (Display *display, XErrorEvent *error) 
{
    if (error->error_code != BadWindow) {
        return old_handler (display, error);
    }
    return 0;
}

GdkFilterReturn
pdfscreenshot_event_filter (GdkXEvent *xevent, GdkEvent *event, gpointer data)
{
    XEvent *ev = (XEvent *) xevent;

    if (ev->type == MapNotify) {
        XTextProperty supported;
        GdkDisplay *display = gdk_x11_lookup_xdisplay(ev->xmap.display);

        XStringListToTextProperty(&supported_str, 1, &supported);

        if (display)
            gdk_x11_display_error_trap_push(display);
        else
            gdk_error_trap_push();
        XSetTextProperty(ev->xmap.display,
            ev->xmap.window,
            &supported,
            gdk_x11_atom_to_xatom(pdfscreenshot_atom));
        if (display)
            gdk_x11_display_error_trap_pop_ignored(display);
        else
            gdk_error_trap_pop_ignored();
    } else if (ev->type == ClientMessage &&
            ev->xclient.message_type == gdk_x11_atom_to_xatom(pdfscreenshot_atom)) {
        if (event->any.window != NULL) {
            GtkWindow *gwin;
            gdk_window_get_user_data(event->any.window, (gpointer *)  &gwin);
            //printf("Taking shot of XWindow 0x%lx, GtkWindow %p\n", ev->xclient.window, gwin);
            g_idle_add(pdfscreenshot_take_shot_soon, gwin);
        } else {
            g_warning("Got a GTK_VECTOR_SCREENSHOT XClientMessage, but window 0x%lx is not known to me.", ev->xclient.window);
        }
    }

    return GDK_FILTER_CONTINUE;
}


/*
 * Module setup function
 */
int
gtk_module_init(gint argc, char *argv[])
{
    filename = NULL;
    type = "pdf";

    pdfscreenshot_atom = gdk_atom_intern ("GTK_VECTOR_SCREENSHOT", FALSE);

    // pdfscreenshot_window_create();

    gdk_window_add_filter (NULL, pdfscreenshot_event_filter, NULL);

    return FALSE;
}




dummy-module/dummy-module.c
/*
 * © 2011 Joachim Breitner <mail@joachim-breitner.de>
 *
 *   This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <gtk/gtk.h>

/*
 * Module setup function
 */
int
gtk_module_init(gint argc, char *argv[])
{
    return FALSE;
}


take-vector-screenshot.c
/*
 * © 2011 Joachim Breitner <mail@joachim-breitner.de>
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */


#include <gtk/gtk.h>
#include <gdk/gdkx.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>

#include <clientwin.h>

GdkAtom pdfscreenshot_atom;
char *supported_str = "supported";

GtkWindow *main_window;

/*
 * This is called once a window has been clicked.
 */
gboolean
pdfscreenshot_window_selected(GtkWidget *grab_window,
                  GdkEventButton *event,
                  GtkButton *button)
{
    // Release the pointer
    gdk_pointer_ungrab(event->time);
    // Un-Press the button (we inhibited the button-release signal in
    // pdfscreenshot_select_window
    gtk_button_released(GTK_BUTTON(button));

    // Use Xlib to find out what window is below the button
    Window selected_window = None;
    int dummy;
    unsigned int dummyU;
    Window dummyW;
    XQueryPointer(gdk_x11_get_default_xdisplay(),
        gdk_x11_get_default_root_xwindow(),
        &dummyW, &selected_window, &dummy, &dummy, &dummy, &dummy, &dummyU);

	// Find the "right" window the same way xwininfo does
    selected_window = Find_Client(gdk_x11_get_default_xdisplay(),
        gdk_x11_get_default_root_xwindow(),
        selected_window);

    if (selected_window != None) {
        // Now we check if the window has the GTK_VECTOR_SCREENSHOT atom set
        XTextProperty supported;
        XGetTextProperty(gdk_x11_get_default_xdisplay(),
            selected_window, 
            &supported,
            gdk_x11_atom_to_xatom(pdfscreenshot_atom));
        if (supported.value != NULL) {
            // Send the GTK_VECTOR_SCREENSHOT ClientMessage to the window. It
            // is important to set xevent.window, as X does not do it by
            // itself.
            XClientMessageEvent xevent;
            xevent.type = ClientMessage;
            xevent.message_type = gdk_x11_atom_to_xatom(pdfscreenshot_atom);
            xevent.format = 32;
            xevent.window  = selected_window;
            XSendEvent(gdk_x11_get_default_xdisplay(), selected_window, 0, 0, (XEvent *)&xevent);
            return TRUE;
        }
    }

    // If we reach here, then the window does not support taking vector screenshots.
    GtkWidget *dialog = gtk_message_dialog_new (main_window,
         GTK_DIALOG_DESTROY_WITH_PARENT,
         GTK_MESSAGE_ERROR,
         GTK_BUTTONS_CLOSE,
         "The selected window does not support taking vector screenshots. Is it "
         "a gtk-3 based application, and did you load the gtk-vector-screenshot "
         "module?");
    gtk_dialog_run (GTK_DIALOG (dialog));
    gtk_widget_destroy (dialog);

    return TRUE;
}


/*
 * Called when the main button is pressed.
 */
gboolean
pdfscreenshot_select_window(GtkWidget *button, GdkEvent *event, GtkWidget *grab_window)
{
    GdkEventMask events = GDK_BUTTON_PRESS_MASK | GDK_BUTTON_RELEASE_MASK;

    // Create a crosshair cursor
    GdkCursor *cursor = gdk_cursor_new_for_display(
        gtk_widget_get_display(button), GDK_CROSSHAIR);
    // And grab the input device, so that all further events are passed to the
    // grab_window
    gdk_device_grab(
        gdk_event_get_device(event),
        gtk_widget_get_window(grab_window),
        GDK_OWNERSHIP_APPLICATION,
        FALSE,
        events,
        cursor,
        event->button.time);

    gdk_cursor_unref(cursor);

    return TRUE;
}


/*
 * The main window, with the one button.
 */
void
pdfscreenshot_window_create()
{
    // Is that icon always installed or do we have to ship it?
    GtkWidget *icon = gtk_image_new_from_icon_name ("camera",GTK_ICON_SIZE_BUTTON);
    gtk_image_set_pixel_size(GTK_IMAGE(icon), 128);

    GtkWidget *button = gtk_button_new_with_label("Take vector screenshot...");
    gtk_button_set_image (GTK_BUTTON(button), icon);
    gtk_button_set_image_position (GTK_BUTTON(button), GTK_POS_TOP);
    gtk_widget_set_tooltip_text (button,
        "Click this button and then an application window to take "
        "a vector screenshot of it.");

    GtkWidget *window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_window_set_icon_name(GTK_WINDOW(window),"camera");
    gtk_window_set_title(GTK_WINDOW(window),"Vector screenshot taker");
    gtk_window_set_resizable(GTK_WINDOW(window), FALSE);
    gtk_window_set_keep_above(GTK_WINDOW(window),TRUE);
    gtk_container_add(GTK_CONTAINER(window), button);
    gtk_widget_show_all(GTK_WIDGET(window));
    main_window = window;

    // Create a dummy window for the pointer grab
    // (maybe try to remove this later)
    GdkEventMask events = GDK_BUTTON_PRESS_MASK | GDK_BUTTON_RELEASE_MASK | GDK_POINTER_MOTION_MASK;

    GtkWidget *grab_window = gtk_window_new(GTK_WINDOW_POPUP);
    gtk_widget_show(grab_window);
    gtk_window_resize(GTK_WINDOW(grab_window), 1, 1);
    gtk_window_move(GTK_WINDOW(grab_window), -100, -100);
    gtk_widget_add_events(grab_window, events);

    g_signal_connect(G_OBJECT(grab_window), "button_release_event",
        G_CALLBACK(pdfscreenshot_window_selected), button);
    g_signal_connect(G_OBJECT(button), "button_release_event",
        G_CALLBACK(pdfscreenshot_select_window), grab_window);

}

int
main(gint argc, char *argv[])
{
    gtk_init(&argc, &argv);

    pdfscreenshot_atom = gdk_atom_intern ("GTK_VECTOR_SCREENSHOT", FALSE);
    pdfscreenshot_window_create();

    gtk_main();

    return 0;
}

